### ISSUE TITLE
Replace reviewer guide with automation.

### ISSUE THREAD

**Issue by hexylena** [galaxyproject/tools-iuc#4349]
Half the things on this list could be checked automatically: https://github.com/galaxyproject/tools-iuc/blob/master/docs/guide_for_reviewers.md

I think this would make a good paper-cut, these checks can be written easily and incrementally and added as needed.

Here are some examples of things that can be checked automatically:

- [ ] Is there a .shed.yml file?
- [ ] Is there a tool .xml file?
- [ ] No tool_dependencies.xml file. (Has been deprecated and is no longer "Best Practice")
- [ ]  Is there a correctly-formatted .shed.yml file?
- [ ]  Does the name match the folder name and, in the case of a single tool, the tool .xml file name? Alphanumeric and underscore _ only, no - (and apparently not `.`)
- [ ]  Is there a description? Are there homepage_url and remote_repository_url fields?
- [ ] Does the tool pass planemo/travis lint with no warnings or errors?
- [ ] 4 spaces indentation?
- [ ] Are the XML elements in the order suggested in the Best Practices Coding Style?
- [ ]  Are the id and name ~sensible and~ not previously used? - I don't even know how you'd check this. Is there a TS api for this?
- [ ]  Does the version follow PEP 440 with +galaxyN?
- [ ] Is there a @TOOL_VERSION@ macro token used
- [ ] Is there a description tag?
- [ ] Is the description of suitable length?
- [ ]  If there is more than one tool present (tool collection), is there a macros.xml file?
- [ ] Is there a <stdio /> element, or does <command /> have a detect_errors attribute, or does the tool specify a profile attribute?
- [ ] Is there a version command?
- [ ] Is it book-ended with <![CDATA[ ... ]]> tags?

and so on, the fast majority of things on the list can be checked automatically and guidance provided in the form of either suggestions or comments directly on the code by a bot, saving reviewers a lot of time and an absolutely too-long checklist for reviewing tools which is a high burden for reviewers and might make it easier to merge things more freely.

**Comment by bernt-matthias** [comment]
- [ ] check for TODO in shed.yml

**Comment by bebatut** [comment]
It would be still nice to have this checklist somewhere readable. I use it for myself (when it has been a while I did not develop a tool) and also I share it with people that start developing tools

**Comment by hexylena** [comment]
If you have a tool that you can run locally, that does the checking automatically, tells you what is wrong (and maybe how to fix), then, is it still needed @bebatut?

**Comment by bernt-matthias** [comment]
- [ ] lower case tool id in shed.yml

**Comment by hexylena** [comment]
The [GTN did this](https://github.com/galaxyproject/training-material/blob/main/bin/lint.rb) and it's been a great success. Feel free to copy from `lint.rb` the syntax for generation rdjson, which can be passed to [`reviewdog`](https://github.com/reviewdog/reviewdog), and [can automatically comment on PRs.](https://github.com/galaxyproject/training-material/blob/main/.github/workflows/ci-main.yml#L62)

it's made the review process significantly nicer for us.

**Comment by hexylena** [comment]
I'm now reviewing a PR and this is annoying enough to do by hand that I'm volunteering for this issue. I'll start by emulating what we do in the GTN. it's 2023 and I'm checking .shed.yml by hand which feels silly.

**Comment by bernt-matthias** [comment]
Still wondering if we should keep the repo / tool linting capabilities at planemo / galaxy/tool-util?

**Comment by hexylena** [comment]
I mean, we could, but I'm not enthusiastic about it. It makes sense overall, e.g. also for the "please follow the IUC model for your own repos" that it's something in planemo, but it means a significantly longer dev cycle for me in the first phase.

I would instead propose:
- we develop the linting here in a single purpose python script, giving us a quick "development" experience + feedback cycle from IUC reviewers.
- once we've got all of the boxes checked, and everyone on board with how it functions, we upstream to planemo.

**Comment by hexylena** [comment]
> lower case tool id in shed.yml

ah, yes, just encountered this one and had to grep to check if upper case was used. Such a great example of a quick and easy technical check. Thanks for adding to this list @bernt-matthias

**Comment by bernt-matthias** [comment]
> but it means a significantly longer dev cycle for me

100% .. 

plus I sometimes have the impression that planemo serves to many tasks.

**Comment by bernt-matthias** [comment]
Maybe you could have a look at this: https://github.com/galaxyproject/tools-iuc/pull/5260 .. sits around for to long without any comments ...

**Comment by hexylena** [comment]
```
$ ag WRAPPER_VERSION -c | cut -f 2 -d : | sum-lines
460
$ ag TOOL_VERSION -c | cut -f 2 -d : | sum-lines   
1379
```

hmm.

**Comment by bernt-matthias** [comment]
I think for PRs we can insist on `TOOL_VERSION`. Over time this will remove all `WRAPPER_VERSION` .. Or we do one big PR now with a big `'s/WRAPPER_VERSION/TOOL_VERSION/g'`?

**Comment by hexylena** [comment]
Let's do it incrementally. There's no reason to make so much noise for a 'non-change'.

**Comment by hexylena** [comment]
All of these are forbidden:

```
./tools/datamash/datamash-transpose.xml: #set file_size_MB = os.path.getsize(str($in_file)) / (1024 * 1024)
./tools/hicexplorer/macros.xml: #set $temp_dir = os.path.abspath(tempfile.mkdtemp())
./tools/hicexplorer/macros.xml: #silent os.system("ln -s %s %s.bam" % (str($i.bamfile), $temp_input_path))
./tools/hicexplorer/macros.xml: #silent os.system("ln -s %s %s.bam.bai" % (str($i.bamfile.metadata.bam_index), $temp_input_path))
./tools/hicexplorer/macros.xml: #set $temp_dir = os.path.abspath(tempfile.mkdtemp())
./tools/hicexplorer/macros.xml: #silent os.system("ln -s %s %s.bw" % (str($i.bigwigfile), $temp_input_path))
./tools/mothur/remove.seqs.xml:#if $os.lstat(str($accnos)).st_size == 0:
./tools/ena_upload/ena_upload.xml: #if os.path.isfile(os.environ.get('GALAXY_ENA_SECRETS', '')):
```

xref https://github.com/galaxyproject/galaxy/issues/16121#issuecomment-1556868629

**Comment by mvdbeek** [comment]
> ./tools/hicexplorer/macros.xml: #silent os.system("ln -s %s %s.bam.bai" % (str($i.bamfile.metadata.bam_index)

that's fine for instance.

> ./tools/ena_upload/ena_upload.xml: #if os.path.isfile(os.environ.get('GALAXY_ENA_SECRETS', '')):

that is fine

**Comment by wm75** [comment]
> > ./tools/ena_upload/ena_upload.xml: #if os.path.isfile(os.environ.get('GALAXY_ENA_SECRETS', '')):
> 
> that is fine

Is it? How is it guaranteed that the job handler can access that file?

**Comment by wm75** [comment]
> shouldn't the symlink be replaced with a command for it? There shouldn't be a reason to use os.system there, it's just unnecessary cheetah addition.

Think so too, plus I don't understand why it would be safe to symlink from cheetah. Same issue as with any other file access?

**Comment by mvdbeek** [comment]
Well no, you're not reading anything there.

**Comment by wm75** [comment]
Ah, you mean the symlink might not be valid for the job handler, but as long as it doesn't use it, it's ok?

**Comment by mvdbeek** [comment]
Right. There are valid reasons to do file path manipulations in cheetah ... for instance you don't need to worry about making things shell-safe right away.

I don't mind if you want to do more shell where it makes sense, what worries me is that this remains possible in the first place.
